File: leetcode/0032-longestValidParentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        if (s == null || s.length() == 0)
            return 0;
        Stack<Integer> stack = new Stack<>();
        int left = -1, len = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') 
                stack.push(i);
            else {
                if (stack.empty()) 
                    left = i;
                else {
                    stack.pop();
                    if (stack.empty()) {
                        int curLen = i - left;
                        len = Math.max(curLen, len);
                    }                        
                    else {
                        int curLen = i - stack.peek();
                        len = Math.max(curLen, len);
                    }
                }
            }
        }
        return len;
    }
}

File: leetcode/0032-longestValidParentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        if s == None or len(s) == 0:
            return 0
        stack = []
        left, length = -1, 0
        for i in range(len(s)):
            if s[i] == "(":
                stack.append(i)
            else:
                if len(stack) == 0:
                    left = i
                else:
                    stack.pop()
                    if len(stack) == 0:
                        cur_len = i - left
                        length = max(cur_len, length)
                    else:
                        cur_len = i - stack[-1]
                        length = max(cur_len, length)
        return length
